develop/items/custom-enchantment-effects.md

```diff
@@ -0,0 +1,97 @@
+            instance.group(
+                    EnchantmentLevelBasedValue.CODEC.fieldOf("amount").forGetter(EnchantmentEntityEffect::amount)
+            ).apply(instance, EnchantmentEntityEffect::new)
+    );
+
+    @Override
+    public void apply(ServerWorld world, int level, EnchantmentEffectContext context, Entity target, Vec3d pos) {
+        // we'll add code here...
+    }
+
+    @Override
+    public MapCodec<? extends EnchantmentEntityEffect> getCodec() {
+        return CODEC;
+    }
+}
+```
+
+Here the `EnchantmentLevelBasedValue amount` indicates a value scaled to the level of the enchantment. We can use this to modify how effective the enchantment is based on level.
+
+## The `apply()`  Event
+
+The bulk of our code will go into the `apply()` event, which is called when the criteria for your enchantment to work is met. We'll later configure this Effect to be called when an entity is hit, but for now let's write simple code to strike the target with lightning.
+
+```java
+@Override
+public void apply(ServerWorld world, int level, EnchantmentEffectContext context, Entity target, Vec3d pos) {
+    if (target instanceof LivingEntity victim){
+        if (context.owner() != null) {
+            // Spawn lightning strikes at the target based on enchantment level
+            float numStrikes = this.amount.getValue(level);
+            for (float i = 0; i < numStrikes; i++) {
+                BlockPos position = victim.getBlockPos();
+                EntityType.LIGHTNING_BOLT.spawn(world, position, SpawnReason.TRIGGERED);
+            }
+        }
+    }
+}
+```
+
+## Registering the Enchantment Effect {#registering-the-enchantment-effect}
+
+Like every other component of your mod, we'll have to add this EnchantmentEffect to Minecraft's registry. To do so, add a class `ModEnchantmentEffects` (or whatever you want to name it) and a helper method to register the enchantment. Be sure to call the `registerModEnchantmentEffects()` in your main class (FabricDocsReference.java if you are following the docs).
+
+@[code transcludeWith=#entrypoint](@/reference/latest/src/main/java/com/example/docs/enchantment/ModEnchantmentEffects.java)
+
+
+## Creating the Enchantment {#creating-the-enchantment}
+
+Now we have an enchantment effect! Lastly, we'll create an enchantment to apply our custom effect to. This can be done by creating a JSON file like in datapacks, however, in this guide we'll generate the JSON dynamically using Fabric's data generation tools. To start, create a folder `data` and within it create an `EnchantmentGenerator` class; optionally you can keep this class within your `enchantment` folder.
+
+Within this class, we'll first register a new enchantment, and then use the `configure()` method to create our JSON programatically.
+
+@[code transcludeWith=#entrypoint](@\reference\latest\src\main\java\com\example\docs\data\EnchantmentGenerator.java)
```

Comment:
Some weird indents within this reference code block, make sure that the comments used to mark this reference in the code are at the same indentation level.

Reply:
This reference is invalid. Make sure the path is correct.

> Not Found: /opt/build/repo\reference\latest\src\main\java\com\example\docs\data\EnchantmentGenerator.java